tests/test_machine.py

```diff
@@ -1,69 +1,68 @@
 # SPDX-License-Identifier: LGPL-2.1+
 
-import os
+import unittest
 from subprocess import TimeoutExpired
+from sys import exc_info
 
-import pytest
-
-import mkosi.machine as machine
 from mkosi.backend import MkosiException
+from mkosi.machine import Machine
 
 
-pytestmark = [
-    pytest.mark.integration,
-    pytest.mark.parametrize("verb", ["boot", "qemu"]),
-    pytest.mark.skipif(os.getuid() != 0, reason="Must be invoked as root.")
-]
-
-
-def test_simple_run(verb: str) -> None:
-    with machine.Machine([verb]) as m:
-        p = m.run(["echo", "This is a test."])
-        assert "This is a test." == p.stdout.strip("\n")
-
-    assert m.exit_code == 0
-
+class MkosiMachineTest(unittest.TestCase):
+    reboot: bool
+    args: str
+    machine: Machine
 
-def test_wrong_command(verb: str) -> None:
-    # First tests with argument check = True from mkosi.backend.run(), therefore we see if an exception is raised
-    with machine.Machine([verb]) as m:
-        with pytest.raises(MkosiException):
-            m.run(["NonExisting", "Command"])
-        with pytest.raises(MkosiException):
-            m.run(["ls", "NullDirectory"])
+    def __init_subclass__(cls, /, args: str, reboot: bool = True) -> None:
+        cls.reboot = reboot
+        cls.args = args
 
-    assert m.exit_code == 0
+    @classmethod
+    def setUpClass(cls) -> None:
+        cls.machine = Machine([cls.args])
+        cls.machine.__enter__()
 
-    # Second group of tests with check = False to see if stderr and returncode have the expected values
-    with machine.Machine([verb]) as m:
-        result = m.run(["NonExisting", "Command"], check=False)
-        assert result.returncode in (203, 127)
+    @classmethod
+    def tearDownClass(cls) -> None:
+        cls.machine.__exit__(*exc_info())
```

Comment:
exc_info() will always return (None, None, None) here since teardownClass() isn't called inside the exception handler block by unittest.

The slightly hacky way to deal with this is to override the internal _callTestMethod() method of unittest.TestCase so we can save the exception result:

Example:

```
def _callTestMethod(self, method: str) -> None:
    try:
        super()._callTestMethod(method)
    except Exception as e:
        self.exc_info = exc_info()
        raise e
```

And then you can pass `self.exc_info()` to `__exit__()` in `tearDownClass()`.